src/main/java/org/springframework/data/elasticsearch/core/SearchShardStatistics.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2013-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core;
+
+import org.springframework.lang.Nullable;
+
+/**
+ * @author Haibo Liu
+ */
+public record SearchShardStatistics(Number failed, Number successful, Number total, @Nullable Number skipped) {
```

Comment:
wouldn't it make sense to add the list of ShardFailures as well?

Reply:
I found this a little awkward: if `ShardFailure` is added into this class, shall we rewrite `ShardFailure`(and `ErrorCause`, `JsonData map` inside it) of our own? or shall we just expose these classes written by `elasticsearch-java` to our users?

I think spring-data-elasticsearch obeys the first rule and almost all the classes exposed to the users are written by spring-data-elasticsearch itself. However, I also notice that when querying aggregations, the class `co.elastic.clients.elasticsearch._types.aggregations.Aggregate` in `elasticsearch-java` is exposed to the users directly.

So which implementation should we choose?

Reply:
It's awkward, yes. And there are already places where Elasticsearch failures are mapped to corresponding classes from Spring Data Elasticsearch. The ugly thing is, that the Elasticsearch client library for this parts  is autogenerated from a schema definition, but this generation is not able to reuse classes, so they redefine them multiple times.

As for exposing Elasticsearch classes: No, we cannot do that, because not every user is using Elasticsearch. Spring Data Elasticsearch also is the base project for [spring-data-opensearch](https://github.com/opensearch-project/spring-data-opensearch) where basically the low level implementation is changed, but the high level stuff is used from Spring Data Elasticsearch. And we cannot introduce Elasticsearch classes into that code.

In version 4, we had Elasticsearch classes happily over the codebase, but with the refactoring needed for the new client in 5 all the code dependent on the old classes went into the `.erhlc.` package and the new one to `.elc.`. There first was the idea of eventually adding a `.orhlc.` for the opensearch rest high level client, but after a call with people from AWS it was decided that all the client specific code has to go into a separate package, they provide the opensearch part in their own project. And everything outside of this pacakge must be independent of a client library.

As for the aggregations: In the Spring Data Elasticsearch code, only interfaces are returned, the concrete implementations come from the client implementations. The main reason why there is no client independent mapping for aggregations is the shere amount of classes that would be needed to map, the last time I checked there were almost 190 classes in the Elasticsearch package. And that needs to be done in every client implementation.

Reply:
Thanks for the detailed explanation. I've seen the corresponding parts, and luckily at least the error cause is reused in the elasticsearch client schema definition :smiling_face_with_tear: